src/jdk.jdeps/share/classes/com/sun/tools/jdeps/ClassFileReader.java

```diff
@@ -241,42 +224,21 @@ public ClassModel getClassFile(String name) throws IOException {
             return null;
         }
 
-        public Iterable<ClassModel> getClassFiles() throws IOException {
-            final Iterator<ClassModel> iter = new DirectoryIterator();
-            return () -> iter;
-        }
-
-        class DirectoryIterator implements Iterator<ClassModel> {
-            private final List<Path> entries;
-            private int index = 0;
-            DirectoryIterator() throws IOException {
-                List<Path> paths = null;
-                try (Stream<Path> stream = Files.walk(path, Integer.MAX_VALUE)) {
-                    paths = stream.filter(ClassFileReader::isClass).toList();
-
-                }
-                this.entries = paths;
-                this.index = 0;
-            }
-
-            public boolean hasNext() {
-                return index != entries.size();
-            }
-
-            public ClassModel next() {
-                if (!hasNext()) {
-                    throw new NoSuchElementException();
-                }
-                Path path = entries.get(index++);
-                try {
-                    return readClassFile(path);
-                } catch (IOException e) {
-                    throw new ClassFileError(e);
-                }
-            }
-
-            public void remove() {
-                throw new UnsupportedOperationException("Not supported yet.");
+        @Override
+        public void processClassFiles(Consumer<ClassModel> handler) throws IOException {
+            try (Stream<Path> stream = Files.walk(path, Integer.MAX_VALUE)) {
+                stream.filter(ClassFileReader::isClass)
+                      .forEach(e -> {
+                          try {
+                              handler.accept(readClassFile(e));
+                          } catch (ClassFileError ex) {
+                              skipEntry(ex, e.toString());
+                          } catch (IOException ex) {
+                              throw new UncheckedIOException(ex);
```

Comment:
Just to point out this was leading to ClassFileError in the old implementation, new implementation will relay the IOException, which I think is proper.

Reply:
I have updated this to skip as well - if the outer structure is malformed, we should not have reached this far and would have encountered IOException earlier.